amethyst_network/src/simulation/transport/resource.rs

```diff
@@ -0,0 +1,165 @@
+}
+
+impl<S: Socket> SimulationTransportResource<S> {
+    /// Create a new `SimulationTransportResource`
+    pub fn new() -> Self {
+        Self {
+            socket: None,
+            messages: VecDeque::new(),
+        }
+    }
+
+    /// Return a mutable reference to the socket if there is one configured.
+    pub fn get_socket_mut(&mut self) -> Option<&mut S> {
+        self.socket.as_mut()
+    }
+
+    /// Set the bound socket to the `SimulationTransportResource`
+    pub fn set_socket(&mut self, socket: S) {
+        self.socket = Some(socket);
+    }
+
+    /// Drops the socket from the `SimulationTransportResource`
+    pub fn drop_socket(&mut self) {
+        self.socket = None;
+    }
+
+    /// Returns whether or not the `SimulationTransportResource` has a bound socket
+    pub fn has_socket(&self) -> bool {
+        self.socket.is_some()
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent on next sim tick.
+    pub fn send(&mut self, destination: SocketAddr, payload: &[u8]) {
+        self.send_with_requirements(
+            destination,
+            payload,
+            S::default_requirement(),
+            UrgencyRequirement::OnTick,
+        );
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent immediately.
+    pub fn send_immediate(&mut self, destination: SocketAddr, payload: &[u8]) {
+        self.send_with_requirements(
+            destination,
+            payload,
+            S::default_requirement(),
+            UrgencyRequirement::Immediate,
+        );
+    }
+
+    /// Create and queue a `Message` with the specified guarantee
+    pub fn send_with_requirements(
+        &mut self,
+        destination: SocketAddr,
+        payload: &[u8],
+        delivery: DeliveryRequirement,
+        timing: UrgencyRequirement,
+    ) {
+        let message = Message::new(destination, payload, delivery, timing);
+        self.messages.push_back(message);
+    }
+
+    /// Returns true if there are messages enqueued to be sent.
+    pub fn has_messages(&self) -> bool {
+        !self.messages.is_empty()
+    }
+
+    /// Drains the messages queue and returns the drained messages
+    pub fn drain_messages(&mut self, mut filter: impl FnMut(&mut Message) -> bool) -> Vec<Message> {
+        let mut drained = Vec::with_capacity(self.messages.len());
+        let mut i = 0;
+        while i != self.messages.len() {
+            if filter(&mut self.messages[i]) {
+                if let Some(m) = self.messages.remove(i) {
+                    drained.push(m);
+                }
+            } else {
+                i += 1;
+            }
+        }
+        drained
+    }
+}
+
+impl<S: Socket> Default for SimulationTransportResource<S> {
+    fn default() -> Self {
+        panic!(
```

Comment:
So I did this because it seems like specs requires that resources implement `Default`. However, I don't want there to be a default implementation of the resource. Basically, I used this to enforce that the end user must provide an implementation with `.with_resource()`